lib/rewards/server/subscribers/world-drop-subscriber.js

```diff
@@ -0,0 +1,164 @@
+
+        let rewards = [];
+        for (let { reward } of itemRewards) {
+            for (let i = 0; i < reward.dropQuantity; i++) {
+                let newReward = sc.deepMergeProperties({}, reward);
+                newReward['rewardPosition'] = this.getRewardPosition(closerWalkableNodes);
+                newReward['tileIndex'] = this.getTileIndex(newReward.rewardPosition, targetObjectBody.worldWidth);
+                newReward['randomRewardId'] = this.getRandomRewardId();
+                newReward['objectBody'] = await this.createRewardRoomObject(roomScene, targetObjectBody, newReward);
+                if (!newReward['objectBody']) {
+                    Logger.error('No object body for reward ' + newReward.id + ' could be created');
+                    continue;
+                }
+                this.setRewardTimeDisappear(newReward['objectBody'], roomScene);
+                rewards.push(newReward);
+            }
+        }
+        return rewards;
+    }
+
+    static getRandomRewardId()
+    {
+        return `reward[${sc.randomChars(8)}]`;
+    }
+
+    static generateWalkableNodesAround(targetObjectBody)
+    {
+        let { currentCol: col, currentRow: row } = targetObjectBody;
+        let pathfinder = targetObjectBody.getPathFinder();
+        let firstNode = pathfinder.grid.getNodeAt(col, row);
+        let firstWorldPos = this.getWorldPosFromNode(firstNode, pathfinder.world.mapJson);
+        let nodes = [firstWorldPos];
+        for (let i = -1; i <= 1; i++) {
+            for (let j = -1; j <= 1; j++) {
+                let node = pathfinder.grid.getNodeAt(col + i, row + j);
+                if (node.walkable) {
+                    nodes.push(this.getWorldPosFromNode(node, pathfinder.world.mapJson));
+                }
+            }
+        }
+        return nodes;
+    }
+
+    static getWorldPosFromNode(node, mapJson)
+    {
+        let tileW = mapJson.tilewidth,
+            tileH = mapJson.tileheight,
+            halfTileW = tileW / 2,
+            halfTileH = tileH / 2;
+        return {
+            x: node.x * tileW + halfTileW,
+            y: node.y * tileH + halfTileH
+        };
+
+    }
+
+    static hasEnoughFreeNodesForRewards(itemRewards, nodes)
+    {
+        return nodes.length >= itemRewards.length;
+    }
+
+    static getTileIndex(node, worldWidth)
+    {
+        return node.x * worldWidth + node.y;
+    }
+
+    static getRewardPosition(closerWalkableNodes)
+    {
+        return closerWalkableNodes.splice(sc.randomInteger(0, closerWalkableNodes.length - 1), 1)[0];
+    }
+
+    static async createRewardRoomObject(roomScene, targetObjectBody, reward)
+    {
+        let dropObjectData = Reward.createDropObjectData(reward, roomScene.roomId);
+        await roomScene.objectsManager.generateObjectFromObjectData(dropObjectData);
+        return await roomScene.roomWorld.createRoomObjectBody(
+            { name: reward.randomRewardId },
```

Comment:
randomRewardId as the layer.name? oh... I see that's for the objectIndex, I will change the `createRoomObjectBody(layer` to use a more proper name like "objectPrefixKey"